Engine/src/ECS/Systems/BaseSystem.bf

```diff
@@ -0,0 +1,318 @@
+
+		protected virtual void Draw(EntityId entityId, List<BaseComponent> components)
+		{
+
+		}
+
+		protected List<BaseComponent> GetComponentsOfEntity(Entity entity)
+		{
+			List<BaseComponent> entityComponents = ?;
+			EntityToComponents.TryGetValue(entity.Id, out entityComponents);
+			return entityComponents;
+		}
+
+		protected virtual Result<void, InitializationError> Initialize()
+		{
+			if (!IsEnabled || IsInitialized)
+			{
+				return .Err(.AlreadyInitialized);
+			}
+			if (InitializeComponents() == .Err)
+			{
+				return .Err(.Unknown);
+			}
+			IsInitialized = true;
+			return .Ok;
+		}
+
+		protected virtual Result<void> Initialize(BaseComponent component)
+		{
+			component.[Friend]IsInitialized = true;
+			return .Ok;
+		}
+
+		protected Result<void> InitializeComponents()
+		{
+			while (_uninitializedComponents.Count != 0)
+			{
+				let component = _uninitializedComponents.Dequeue();
+				if (!component.IsInitialized)
+				{
+					if (Initialize(component) == .Err)
+					{
+						return .Err;
+					}
+				}
+			}
+			return .Ok;
+		}
+
+		protected virtual void PostUpdate()
+		{
+			ClearEmptyEntities();
+		}
+
+		protected virtual void PreUpdate()
+		{
+			ClearEmptyEntities();
+		}
+
+		protected bool RefreshEntityRegistration(EntityId entityId)
+		{
+			List<BaseComponent> components = ?;
+			if (!EntityToComponents.TryGetValue(entityId, out components))
+			{
+				return false;
+			}
+
+			let componentsToRemove = new List<BaseComponent>();
+			var valid = true;
+			// Check if every required Component is present
+			for (let type in _registeredTypes)
+			{
+				// Check all components for the registered type
+				bool found = false;
+				for (let component in components)
+				{
+					if (CanBeRegisteredAsType(component, type))
+					{
+						found = true;
+						break;
+					}
+				}
+				if (found)
+				{
+					continue;
+				}
+				// Registered type is not present. This invalidates all components that are registered to this entity.
+				for (let component in components)
+				{
+					componentsToRemove.Add(component);
+				}
+				valid = false;
+			}
+			for (let component in componentsToRemove)
+			{
+				RemoveComponent(component);
+			}
+
+			return valid;
+		}
+
+		protected abstract void RegisterComponentTypes();
+
+		protected virtual bool RemoveComponent(BaseComponent component)
+		{
+			let parent = component.Parent;
+			List<BaseComponent> components = ?;
+			if (parent == null || !EntityToComponents.TryGetValue(parent.Id, out components))
+			{
+				return false;
+			}
+			if (components.Remove(component))
+			{
+				_entityRegistrationChecks.Add(parent.Id);
+				return true;
+			}
+			return false;
+		}
+
+		protected virtual void Update(float delta)
+		{
+			if (!IsEnabled)
+			{
+				return;
+			}
+			InitializeComponents();
+
+			// Before running the update, check registration status of all components.
+			for (let entity in _entityRegistrationChecks)
+			{
+				RefreshEntityRegistration(entity);

```

Comment:
This can change content of `_entityRegistrationChecks` while iterating which is not supported.

Reply:
You are right. Thanks for catching that. I thought I had caught those cases.